### PR TITLE
[youtube:comments] Move comment extraction to new API

### LINKED ISSUE #485: Do not attempt to get a YT video's comments even if --get-comments is passed if YT doesn't provide comment continuation token
## Checklist

- [x] I'm reporting a broken site support issue
- [x] I've verified that I'm running yt-dlp version **2021.07.07**
- [ ] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar bug reports including closed ones
- [x] I've read bugs section in FAQ

## Verbose log

```
yt-dlp "https://www.youtube.com/watch?v=j8oHN6Hi6bc" -v --ignore-config --write-comments --no-check-certificate --proxy "http://localhost:8888/"
[debug] Command-line config: ['https://www.youtube.com/watch?v=j8oHN6Hi6bc', '-v', '--ignore-config', '--write-comments', '--no-check-certificate', '--proxy', 'http://localhost:8888/']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, pref cp1252
[debug] yt-dlp version 2021.07.07
[debug] Python version 3.9.1 (CPython 64bit) - Windows-10-10.0.19041-SP0
[debug] exe versions: ffmpeg n4.3.2-163-g6c414cf8f7, ffprobe n4.3.2-163-g6c414cf8f7
[debug] Proxy map: {'http': 'http://localhost:8888/', 'https': 'http://localhost:8888/'}
[debug] [youtube] Extracting URL: https://www.youtube.com/watch?v=j8oHN6Hi6bc
[youtube] j8oHN6Hi6bc: Downloading webpage
[youtube] j8oHN6Hi6bc: Downloading MPD manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, vcodec:vp9.2(10), acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] j8oHN6Hi6bc: Downloading 1 format(s): 244+140
ERROR: 'NoneType' object is not iterable
Traceback (most recent call last):
  File "c:\users\abc\scoop\apps\python\current\lib\site-packages\yt_dlp\YoutubeDL.py", line 1142, in wrapper
    return func(self, *args, **kwargs)
  File "c:\users\abc\scoop\apps\python\current\lib\site-packages\yt_dlp\YoutubeDL.py", line 1178, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "c:\users\abc\scoop\apps\python\current\lib\site-packages\yt_dlp\YoutubeDL.py", line 1219, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "c:\users\abc\scoop\apps\python\current\lib\site-packages\yt_dlp\YoutubeDL.py", line 2198, in process_video_result
    self.process_info(new_info)
  File "c:\users\abc\scoop\apps\python\current\lib\site-packages\yt_dlp\YoutubeDL.py", line 2358, in process_info
    self.post_extract(info_dict)
  File "c:\users\abc\scoop\apps\python\current\lib\site-packages\yt_dlp\YoutubeDL.py", line 2860, in post_extract
    actual_post_extract(info_dict or {})
  File "c:\users\abc\scoop\apps\python\current\lib\site-packages\yt_dlp\YoutubeDL.py", line 2852, in actual_post_extract
    extra = post_extractor().items()
  File "c:\users\abc\scoop\apps\python\current\lib\site-packages\yt_dlp\extractor\youtube.py", line 2927, in <lambda>
    info['__post_extractor'] = lambda: self._extract_comments(ytcfg, video_id, contents, webpage, xsrf_token)
  File "c:\users\abc\scoop\apps\python\current\lib\site-packages\yt_dlp\extractor\youtube.py", line 2145, in _extract_comments
    for entry in contents:
TypeError: 'NoneType' object is not iterable
```

## Description

Reporting a small bug with the youtube extractor

For a few old youtube videos, the channel the video is uploaded to is unavailable. Somehow this causes youtube not to serve the comment continuation token at all. Therefore, if a user were to download one of these broken videos with yt-dlp with the --get-comments flag, yt-dlp throws a TypeError instead of catching the error and continuing. 

Some videos where youtube doesn't serve a comment continuation token:
```
https://www.youtube.com/watch?v=wWhLVn8shzo
https://www.youtube.com/watch?v=D0sbmx7PupM
https://www.youtube.com/watch?v=j8oHN6Hi6bc
https://www.youtube.com/watch?v=SAcdZ7I87gk
https://www.youtube.com/watch?v=Rw8UGdWSwqo
https://www.youtube.com/watch?v=B0TfFjLEbHw
```

ping @colethedj as it may be of interest

### PR DESCRIPTION
**Pull request by colethedj**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Bug fix
- [X] Improvement
- [ ] New extractor
- [X] New feature

---

### Description of your *pull request* and other information

Mostly just a draft at the moment. Likely will not merge until YouTube completely moves over to the new API for comment extraction on web client.

Fixes https://github.com/yt-dlp/yt-dlp/issues/438

### ISSUES:

- YouTube hasn't fully switched over to the new API yet, and still provides the old token. Currently I handle this by manually generating a new API initial token. I also added  handling for the old API JSON structure (as that will still be used if visitorData was linked to the old comment section). 

- YouTube sometimes ends up providing continuation tokens but no comment data. Mostly happens in popular sorting mode, but I have observed this happening for the odd reply thread. Currently I just break if it detects no continuation data, and stopping if that happens for the parent pages.

This will need some good testing to check any edge cases I may have missed.

### TODO:
- [X] Rewrite to support new response structure
- [X] Add support for old response structure
- [X] Convert old initial token to new token
- [X] Documentation (`--extractor-args youtube:comment_sort=top|new`)
- [ ] Testing
- [X] Fix #485
- [X] Implement https://github.com/yt-dlp/yt-dlp/issues/481? - (edit: `--extractor-args youtube:max_comments=xyz;max_comment_depth=1|2`

### CODE REVIEW

**Comment by colethedj** [comment]
YouTube appears to be testing different response structures atm.

New API with old response:
image

New API with new response:
image

Noticed they've completely changed the location of the data we want, so we'll have to rework the relevant extraction code.

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -533,6 +539,74 @@ def _generate_api_headers(self, ytcfg=None, identity_token=None, account_syncid=
             headers['X-Origin'] = origin
         return headers
 
+    @staticmethod
+    def _build_api_continuation_query(continuation, ctp=None):
+        query = {
+            'continuation': continuation
+        }
+        # TODO: clickTrackingParams appears to be contained within the context
+        # which is extracted from ytcfg (and is different from one extracted with continuation)
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:550]: does this cause any issues?

**Review by pukkandan — APPROVED** [review]
LGTM

This seems to correctly handle both old and new tokens now. You can merge if you are finished making changes

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -1941,10 +2017,43 @@ def _extract_comment(self, comment_renderer, parent=None):
...
+                comments_header_renderer = try_get(content, lambda x: x['commentsHeaderRenderer'])
+                expected_comment_count = try_get(comments_header_renderer,
+                                                 (lambda x: x['countText']['runs'][0]['text'],
+                                                  lambda x: x['commentsCount']['runs'][0]['text']),
+                                                 compat_str)
+                if expected_comment_count:
+                    comment_counts[1] = str_to_int(expected_comment_count)
+                    self.to_screen('Downloading ~%d comments' % str_to_int(expected_comment_count))
+                    _total_comments = comment_counts[1]
+                sort_mode_str = self._configuration_arg('comment_sort', [''])[0]
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2035]: oh, I forgot. Make sure to document this

File: yt_dlp/extractor/youtube.py (hunk 1 above)
Comment (colethedj) [reply yt_dlp/extractor/youtube.py:550]: Doesn't appear to at the moment, but might be good to keep in mind.

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -2255,7 +2310,8 @@ def get_text(x):
                 default_client=ytm_client,
                 note='Downloading %sremix player API JSON' % ('mobile ' if force_mobile_client else ''))
 
-            ytm_streaming_data = try_get(ytm_player_response, lambda x: x['streamingData']) or {}
+            if isinstance(ytm_player_response, dict):
+                ytm_streaming_data = ytm_player_response.get('streamingData') or {}
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2314]: I hadn't thought about `ytm_player_response` not being a dict. In that case, you can just revert my changes and go back to:
```suggestion
           # ytm_player_response need not be a dict
            ytm_streaming_data = try_get(ytm_player_response, lambda x: x['streamingData']) or {}
```

but 
```py
            ytm_player_response = self._extract_response(...) or {}
            ytm_streaming_data = ytm_player_response.get('streamingData') or {}
```
is more clearer

File: yt_dlp/extractor/youtube.py (hunk 1 above)
Comment (pukkandan) [reply yt_dlp/extractor/youtube.py]: > but might be good to keep in mind.

Then why'd you remove the comment? 😕

Comment (colethedj) [reply yt_dlp/extractor/youtube.py]: good question. I added a comment back

File: README.md [hunk 4]
```diff
@@ -1336,7 +1336,8 @@ The following extractors use this feature:
     * `player_client`: `web` (default) or `android` (force use the android client fallbacks for video extraction)
     * `player_skip`: `configs` - skip requests if applicable for client configs and use defaults
     * `comment_sort`: `popular` or `newest` (default) - choose comment sorting mode on YouTube's side.
-
+    * `max_comments`: Maximum amount of comments to download (default all).
+    * `max_comment_depth`: maximum depth for nested comments. YouTube supports depths 1 or 2 (default).
```
Comment (pukkandan) [inline README.md:1340]: oh, I totally forgot that, lol. That makes this option pretty useless. Anyway, since it is now implemented, lets keep it

**Review by pukkandan — APPROVED** [review]
